Task/dashboardDesign

jduong09 opened the pull request:
Resolves #35

ziggysauce commented on client/public/index.html:
```diff
@@ -24,8 +24,10 @@
       work correctly both with client-side routing and a non-root public URL.
       Learn how to configure a non-root public URL by running `npm run build`.
     -->
+    <!-- 
     <link href="https://cdn.jsdelivr.net/npm/bootstrap@5.0.2/dist/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-EVSTQN3/azprG1Anm3QDgpJLIm9Nao0Yz1ztcQTwFspd3yD65VohhpuuCOmLASjC" crossorigin="anonymous">
     <script src="https://cdn.jsdelivr.net/npm/bootstrap@5.0.2/dist/js/bootstrap.bundle.min.js" integrity="sha384-MrcW6ZMFYlzcLA8Nl+NtUVF0sA7MsXsP1UyJoMp4YLEuNSfAP+JcXn/tWtIaxVXM" crossorigin="anonymous"></script>
+    -->
```
we can remove this since we're using the downloaded version

ziggysauce commented on client/src/components/date/dashboardCalendar.js:
```diff
@@ -0,0 +1,17 @@
+
+const DashboardCalendar = ({ subscriptions }) => {
+  const dueDates = [];
+  subscriptions.map((subscription) => {
+    const { dueDate } = subscription;
+    dueDates.push(new Date(dueDate.nextDueDate));
```
if this is the only data used you can just destructure directly:

```suggestion
  subscriptions.map(({ dueDate }) => {
    dueDates.push(new Date(dueDate.nextDueDate));
```

ziggysauce commented on client/src/components/date/dashboardCalendar.js:
```diff
@@ -0,0 +1,17 @@
+import DayPicker from 'react-day-picker';
+
+const DashboardCalendar = ({ subscriptions }) => {
+  const dueDates = [];
+  subscriptions.map((subscription) => {
+    const { dueDate } = subscription;
+    dueDates.push(new Date(dueDate.nextDueDate));
+    return 'hi';
+  });
```
actually, if you're just tryna get an array of dueDates and want to use `.map()` you don't need to create and array and push into it. instead you should just do:

```suggestion
  const dueDates = subscriptions.map(({ dueDate }) => new Date(dueDate.nextDueDate));
```

ziggysauce commented on client/src/components/landingPage.js:
```diff
@@ -1,20 +1,34 @@
-import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
-import { faSignInAlt } from '@fortawesome/free-solid-svg-icons';
+import logo from '../assets/watering-can.png';
+
```
i thought you got rid of this line for me?

ziggysauce commented on client/src/components/user/dashboard.js:
```diff
@@ -1,39 +1,280 @@
+                  type="button"
+                  onClick={() => this.setAddingSubscription(true)}
+                >
+                  +Create
```
```suggestion
                  + Create
```

ziggysauce commented on client/src/components/user/dashboard.js:
```diff
@@ -1,39 +1,280 @@
+              setActiveSubscription={this.setActiveSubscription}
+              handleDelete={this.handleDelete}
+            />
+            <div className="col">
```
I would add a little margin above here so that the dotted line isn't squished up against the existing subscription list

```suggestion
            <div className="col mt-2">
```

ziggysauce commented on client/src/components/subscription/subscriptionForm.js:
```diff
@@ -124,103 +126,142 @@ class SubscriptionForm extends React.Component {
+            <div className="col-4 col-md-3">Repeat:</div>
+            <div className="col-8 col-md-7">
+              <select className="form-control" id="due-date-select" value={frequency} onChange={(event) => this.handleChange(event, 'frequency')} required >
+                <option value="" defaultValue>--Please Choose an Option--</option>
```
This text can get cutoff in tablet view, let's just shorten it for simplicity. I'd also add the `disabled` attribute

```suggestion
                <option value="" defaultValue disabled>-- Select --</option>
```

ziggysauce commented on client/src/components/subscription/subscriptionListCard.js:
```diff
@@ -0,0 +1,41 @@
+  return (
+    <div className="w-100 p-2 d-flex justify-content-between align-items-center">
+      <button className="btn" type="button" onClick={() => setActiveSubscription(details)}>
+        <ul className="text-start">
```
The text looks misaligned and provides too much white space on the left due to the padding from the `<ul>` element

```suggestion
        <ul className="text-start px-0">
```

ziggysauce commented on client/src/components/subscription/subscriptionForm.js:
```diff
@@ -124,103 +126,142 @@ class SubscriptionForm extends React.Component {
+              <label className="col d-flex form-label" htmlFor="occurrence">
+                <div className="col-4 col-md-3">Every (?) Weeks:</div>
+                <div className="col-8 col-md-7">
+                <select className="form-control" id="occurrence" placeholder="ex: Every 2 weeks" value={occurrence} onChange={(event) => this.handleChange(event, 'occurrence')}>
```
The dropdown doesn't look like a select dropdown and instead looks like a text input field b/c it doesn't have the little arrow on the right side. I'd suggest using the `form-select` bootstrap class for this instead of the `form-control` that you're currently using

```suggestion
                <select className="form-select" id="occurrence" placeholder="ex: Every 2 weeks" value={occurrence} onChange={(event) => this.handleChange(event, 'occurrence')}>
```

ziggysauce commented on client/src/components/subscription/subscriptionForm.js:
```diff
@@ -124,103 +126,142 @@ class SubscriptionForm extends React.Component {
+          <label className="col d-flex form-label align-items-center col-md-6" htmlFor="due-date-select">
+            <div className="col-4 col-md-3">Repeat:</div>
+            <div className="col-8 col-md-7">
+              <select className="form-control" id="due-date-select" value={frequency} onChange={(event) => this.handleChange(event, 'frequency')} required >
```
The dropdown doesn't look like a select dropdown and instead looks like a text input field b/c it doesn't have the little arrow on the right side. I'd suggest using the `form-select` bootstrap class for this instead of the `form-control` that you're currently using

```suggestion
              <select className="form-select" id="due-date-select" value={frequency} onChange={(event) => this.handleChange(event, 'frequency')} required >
```

ziggysauce reviewed (changes requested):
Awesome work so far with this UI update! It's looking WAYYYY better already. Def. some work still needs to be done, but this is a really great start so far. My first pass review is primarily ui/ux. I will go through the code a bit deeper for syntax, logic, and optimizations after you've made changes.

- When I try to click on the notifications button I get hit with some code error 
  - ~Also the profile pic img isn't working~ nvm it loaded for me after refreshing. but we should give this some border-radius, or use bootstrap's `rounded` class, or perhaps make it a circle
  - Also give some padding between the sign out button and the icon
<img width="192" alt="Screen Shot 2022-03-13 at 11 12 00 PM" src="https://user-images.githubusercontent.com/25065329/158098497-ef06c0c0-27f6-415f-9411-1abed19a757d.png">
<img width="1431" alt="Screen Shot 2022-03-13 at 11 11 35 PM" src="https://user-images.githubusercontent.com/25065329/158098513-a7d2f903-b9ea-4139-baa9-cf5b72306b55.png">

- Right now you have to click the specific area around the subscription in order to see the preview in the calendar, as opposed to clicking anywhere on the card. I would either (1) make the entire card clickable instead or (2) make it clear that you have to click on the specific area of the subscription since that's not obvious right now

<img width="472" alt="Screen Shot 2022-03-13 at 11 13 28 PM" src="https://user-images.githubusercontent.com/25065329/158098688-7df7b417-65d4-45f3-bb5d-a25f8582bfc0.png">

- Right now you have an infinite loop of rendering which is causing a major memory leak. It looks like it's from the `useEffect()` in `dashboardCalendar.js`:

https://user-images.githubusercontent.com/25065329/158100380-9f541345-1524-4ee0-afd5-fcb14b8d5c60.mov

^^ Make sure to often check your dev console to look for errors such as this.

Right now you can provide a solution by just doing a strict deep copy array check:
```javascript
  useEffect(() => {
    const dates = subscriptions.map((subscription) => new Date(subscription.dueDate.nextDueDate));
    if(JSON.stringify(dueDates) !== JSON.stringify(dates)) {
      setDueDates(dates);
    }
  }, [dueDates, subscriptions]);
```
however this isn't always the best practice. So I'd do some googling to find out if there's a better solution. But this will at least stop the infinite loop until then.


- This feels like a ton of unused white space. Don't know what the solution is yet, but for the time being I think vertically centering the subscription info instead of having it at the top might make better use of this space so it feels less empty:

<img width="764" alt="Screen Shot 2022-03-13 at 11 37 14 PM" src="https://user-images.githubusercontent.com/25065329/158100712-a4c27c47-f6f9-4ed4-8f3f-fb8fcfc5cf6e.png">

- Don't know if it makes a whole lot of sense to label this button as "Home". At least in desktop, since right now the entire page is still visible, the button doesn't actually navigate you anywhere, it simply hides the current subscription and displays the full calendar view. "Close" may be a better term here

<img width="1178" alt="Screen Shot 2022-03-13 at 11 39 18 PM" src="https://user-images.githubusercontent.com/25065329/158100897-081c96d4-6abf-404a-822d-c51cde347b42.png">

- Couple questions about this one. (1) What happened to displaying the text that was like "3/15/22, Monthly" to indicate that the repeated cadence is monthly as opposed to just showing the due date? And (2) we probs don't need to show the "Days Selected" text unless the user actually selects one or more dates

<img width="1179" alt="Screen Shot 2022-03-13 at 11 41 50 PM" src="https://user-images.githubusercontent.com/25065329/158101351-571932d3-0b36-41db-8f00-850b826106b4.png">

- When I click on a subscription, it shows me the enlarged sub view, but when I go to click the "edit" button, I don't see the editing UI until I close the enlarged sub view. But we should make it so that whatever view the user clicks on is what they see. Thus, if they're in the middle of editing and click the enlarged view we may want to have some warning modal that says "Unsaved changes, are you sure you want to exit the edit mode" or something of that nature. Similarly, when I'm viewing a sub, then click the "+Create" to add one, I can't do so until closing the current sub.

- Here I've created a sub with a monthly cadence and two dates (4/1 and 4/15) but only 4/1 shows up despite it saving correctly in the database with both dates:

<img width="756" alt="Screen Shot 2022-03-13 at 11 54 18 PM" src="https://user-images.githubusercontent.com/25065329/158102208-cc121278-a8df-436c-ac0a-ed711ecca085.png">

- In mobile, the subscriptions should be center aligned:

<img width="1170" alt="Screen Shot 2022-03-14 at 12 01 04 AM" src="https://user-images.githubusercontent.com/25065329/158102775-6d6be052-18c1-44d8-b40a-f8904b90a2bb.png">

- On mobile, the drawer's horizontal content is slightly cutoff:
<img width="1182" alt="Screen Shot 2022-03-14 at 12 04 46 AM" src="https://user-images.githubusercontent.com/25065329/158103028-7fb268ae-8f94-484d-9707-395448605faa.png">. 

Couple things to address here as well:
  - I'd get rid of the purple top border. Doesn't seem necessary and looks cleaner when the drawer is just white
  - I'd lessen the border radius on top. 3em is quite a bit, and it'll look better imo with just 1em
  - The height of the drawer probably only needs to be 75vh instead of 95vh

jduong09 commented:
### Fixes
- Memory leak that was being caused by useEffect function in dashboardCalendar.
- Error caused by clicking 'Notifications' in PFP dropdown.
- SubscriptionDetail calendar not showing all due dates.
- On mobile, calendar slightly cutoff on forms.
- Issue where user has to close one view, in order to see another view.
- Removed unnecessary styling in App.css

### Changes/Additions
- Added 'edit' and 'delete' buttons to Subscription Detail view.
- Matched header on dashboard with landing page.
- Added styling to subscription list.
- Aligned create/update subscription forms.
- Added responsive font-size to calendar.
- Removed border color on mobile create/subscription drawer.

jduong09 commented:
### Changes/Additions
- Fixed line separator on mobile view between main component and subscriptions list.
- Made Calendar font larger to fill empty space.
- Changed font to use Noto Sans.
- Decreased height for the highlight on landing page header.
- Changed toast notifications css to match overall web app.
- Changed header 1 font-weight from 300 to 400.

ziggysauce commented on client/src/components/subscription/subscriptionDetail.js:
```diff
@@ -0,0 +1,85 @@
+    setEditingSubscription(details);
+  }, [setActiveSubscription, setEditingSubscription, details]);
+
+  const clickDelete = useCallback((e) => {
+    e.stopPropagation();
+    setActiveSubscription(false);
+    handleDelete(subscriptionUuid);
+  }, [setActiveSubscription, handleDelete, subscriptionUuid]);
```
we can handle this in the future, but I think having a confirmation modal (Are you sure you want to delete this?) would be good here

jduong09 replied:
Implemented in the task/addModal PR.

ziggysauce commented on client/src/components/user/dashboard.js:
```diff
@@ -1,39 +1,265 @@
-            Sign Out!
-            <FontAwesomeIcon icon={faSignOutAlt} />
+      <div className="h-100 d-flex flex-column">
+        <header className="navbar p-2 d-flex justify-content-between align-items-center text-dark border-bottom shadow-sm">
```
i think this could use a bit more horizontal padding on desktop

```suggestion
        <header className="navbar py-2 px-4 d-flex justify-content-between align-items-center text-dark border-bottom shadow-sm">
```

ziggysauce replied:
And on mobile, we can maybe shrink the header size so that the icon, date, and pfp all fit in one line since right now it wraps:
<img width="448" alt="Screen Shot 2022-03-23 at 1 11 50 AM" src="https://user-images.githubusercontent.com/25065329/159628398-e576f04e-d251-41b6-8a5b-de2163eb1fdc.png">
 
But if you did say `h3` as the class for the header it would look like this:
<img width="448" alt="Screen Shot 2022-03-23 at 1 13 21 AM" src="https://user-images.githubusercontent.com/25065329/159628561-cae313f5-1d1d-4a33-8834-19fa38917bc5.png">

ziggysauce commented on client/src/components/user/dashboard.js:
```diff
@@ -1,39 +1,265 @@
+            <img src={logo} alt="wateringCanIcon" height="60" />
+            <div className="align-self-center d-none d-md-block">Water Your Subs</div>
           </a>
+          <h1 className="display-4 fw-normal">{`${new Date().toDateString()}`}</h1>
```
we should modify this a bit to have a comma between the day and year. right now it looks like `Wed March 23 2022` but it should look like `Wed March 23, 2022`

ziggysauce commented on client/src/custom.scss:
```diff
@@ -0,0 +1,67 @@
+  position: relative !important;
+}
+
+.innerButtonEdit {
+  position: absolute !important;
+  right: 0 !important;
+  top: 0 !important;
+}
+
+.innerButtonDelete {
+  position: absolute !important;
+  right: 0 !important;
+  bottom: 0 !important;
+}
```
I think instead of absolutely positioning those icon buttons, we can achieve the same effect just using the flexbox classes in bootstrap

if you have a parent1 container with a child left container (name, due date, amount) and a child right container (edit and delete icons), then the parent just needs `d-flex justify-content-between` to have them be on opposite sides of the parent container. then within the child right container, if you vertically align them with `d-flex flex-column` it should stack vertically, then you can justify them either between or around or however you think looks best

ziggysauce reviewed (changes requested):
<img width="636" alt="Screen Shot 2022-03-23 at 1 22 20 AM" src="https://user-images.githubusercontent.com/25065329/159629545-ae56fc58-86c8-4541-a12f-41e42e181570.png">

I think having the "close" button and the "x" button in the top right corner are kinda redundant. probs just need one and not both

jduong09 commented:
### Changes/Additions
- Removed specific classes for positioning icons in subscription list card. Used bootstrap flex instead.
- Adding vertical padding to elements in header.
- Made h1 tag display as h3, to fit for mobile devices better.
- Designed PFP in header to be a circle, instead of the previous rectangle.
- Removed close button from subscription detail, added 'x' close in top right.
- Manipulated todays date to display correctly.

ziggysauce commented on client/src/components/user/dashboard.js:
```diff
@@ -198,15 +204,15 @@ class Dashboard extends React.Component {
             <div className="align-self-center d-none d-md-block">Water Your Subs</div>
           </a>
-          <h1 className="display-4 fw-normal">{`${new Date().toDateString()}`}</h1>
+          <h1 className="h3 fw-normal">{concatenatedString}</h1>
```
this makes the header always smaller, i was suggesting to only make it smaller when in mobile view

ziggysauce reviewed (changes requested):
on mobile, instead of labels on the left and inputs on the right for adding/editing a subscription i think it would make more sense to just stack it vertically, so that it goes

```
Label
Input
```

such as

```
Name
[.                 ].  <-- text input here
```
<img width="340" alt="Screen Shot 2022-03-24 at 10 44 52 PM" src="https://user-images.githubusercontent.com/25065329/160044607-3db93c03-71a1-46ed-b8d3-a839d7f9f655.png">

jduong09 commented:
### Changes/Additions
- Made label and input change that Dan requested on mobile.
- Added css to Calendar on subscription form.
- Made some layout changes to create/update forms on desktop.
- Added spacing to list card. Removed purple hover.

### Things To Do
- Highlight active subscription that user is viewing on the subscription list.
- Keep offcanvas (drawer) up when user submits without selecting a date.
- Make sure subscription detail view is similar to create/update.
- Design subscription detail view.
- Late toast notification is ugly.

jduong09 commented:
### Changes/Additions
- Made changes to subscription detail calendar to be similar to create/update calendar.
- Changed layout of subscription detail to fit more nicely on mobile view.
- Changed headings for all main component view, to be similar to each other and the subscription list.
- Removed line separator, added border instead.
- Changed edit and delete buttons on subscription detail to be similar to create button.
- Made header in nav bar primary color, and added some font-weight to make it stand out.

### To-Do
- Design main calendar in first view to better fit it's container.
- Change dropdown on PFP padding.
- Change logo, right now it is clickable although it doesn't need to be.

ziggysauce commented on server/api/routes/users.js:
```diff
@@ -16,4 +16,11 @@ router.route('/')
+  const { user }  = req.session.passport;
+  res.status(200).json({ name: user.displayName, pfp: user.picture });
+});
+
+
+
```
get rid of some of these extra empty lines

ziggysauce commented on client/src/components/user/dashboard.js:
```diff
@@ -1,39 +1,406 @@
+const todaysDate = new Date().toDateString();
+const year = todaysDate.slice(todaysDate.length - 4);
+const monthAndDay = todaysDate.slice(0, todaysDate.length - 5);
+const concatenatedString = `${monthAndDay}, ${year}`;
+
+
```
get rid of extra empty line
